src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -2006,12 +2017,71 @@ class Point
+            var expected = new[]
             {
-                return node.Name;
+                new { Source = "Point(3, string { Length: 2 })", Type = "System.Object", ConvertedType = "Point" },
+                new { Source = "3", Type = "System.Object", ConvertedType = "System.Int32" },
+                new { Source = "string { Length: 2 }", Type = "System.Object", ConvertedType = "System.String" },
+                new { Source = "2", Type = "System.Int32", ConvertedType = "System.Int32" },
+                new { Source = "(_, { })", Type = "Point", ConvertedType = "Point" },
+                new { Source = "_", Type = "System.Object", ConvertedType = "System.Object" },
+                new { Source = "{ }", Type = "System.Object", ConvertedType = "System.Object" },
+                new { Source = "Point({ }, { }, { })", Type = "Point", ConvertedType = "Point" },
+                new { Source = "{ }", Type = "?", ConvertedType = "?" },
+                new { Source = "{ }", Type = "?", ConvertedType = "?" },
+                new { Source = "{ }", Type = "?", ConvertedType = "?" },
+                new { Source = "Point(, { })", Type = "Point", ConvertedType = "Point" },
+                new { Source = "", Type = "System.Object", ConvertedType = "System.Object" },
+                new { Source = "{ }", Type = "System.Object", ConvertedType = "System.Object" },
+            };
+            var compilation = CreateCompilation(source, options: TestOptions.ReleaseDll);
+            compilation.VerifyDiagnostics(
+                // (10,24): error CS8400: Pattern missing
+                //         if (p is Point(, { })) { }
+                Diagnostic(ErrorCode.ERR_MissingPattern, ",").WithLocation(10, 24),
+                // (9,18): error CS1501: No overload for method 'Deconstruct' takes 3 arguments
+                //         if (p is Point({ }, { }, { })) { }
+                Diagnostic(ErrorCode.ERR_BadArgCount, "Point({ }, { }, { })").WithArguments("Deconstruct", "3").WithLocation(9, 18),
+                // (9,18): error CS8129: No suitable Deconstruct instance or extension method was found for type 'Point', with 3 out parameters and a void return type.
+                //         if (p is Point({ }, { }, { })) { }
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "Point({ }, { }, { })").WithArguments("Point", "3").WithLocation(9, 18)
+                );
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+            int i = 0;
+            foreach (var pat in tree.GetRoot().DescendantNodesAndSelf().OfType<PatternSyntax>())
+            {
+                var typeInfo = model.GetTypeInfo(pat);
+                var ex = expected[i++];
+                Assert.Equal(ex.Source, pat.ToString());
+                Assert.Equal(ex.Type, typeInfo.Type.ToTestDisplayString());
+                Assert.Equal(ex.ConvertedType, typeInfo.ConvertedType.ToTestDisplayString());
```

Comment:
@gafter Another general discussion: could we reconsider factoring `PropertyPatternSyntax` (with a type, property sub-pattern and designation) and `DeconstructionPatternSyntax` (with the same plus parens and list of patterns)?

The parens and list of patterns are pulled into a deconstruction sub-pattern. Then we can have one node with optional type, optional positional sub-pattern and optional property sub-pattern.

I understand that you wanted to have the syntax nodes enforce some constraints (you can't be missing the type, positional and property sub-patterns). 

But on the other hand, my experience dealing with the syntax API so far is that it would be easier to deal with and reason about.

Reply:
what about exposing their common properties through an abstract base class (Type, PropertySubpattern, Designation)?

Reply:
@Neme12 See https://github.com/dotnet/roslyn/issues/26540